File: Map/src/main/python/solution697.py
# https://leetcode-cn.com/problems/degree-of-an-array/

from typing import List


class Solution:
    def findShortestSubArray(self, nums: List[int]) -> int:
        cnt_map = {}
        max_count = 0
        res = len(nums)

        for i in range(len(nums)):
            if cnt_map.get(nums[i]):
                cnt_map[nums[i]][1] += 1
            else:
                cnt_map[nums[i]] = [i, 1]

            if cnt_map[nums[i]][1] > max_count:
                max_count = cnt_map[nums[i]][1]

                res = i - cnt_map[nums[i]][0] + 1

            if cnt_map[nums[i]][1] == max_count:
                res = min(res, i - cnt_map[nums[i]][0] + 1)

        return res


s = Solution()
nums = [1, 2, 2, 3, 1]
res = s.findShortestSubArray(nums)

print(res)

File: Map/src/main/java/Solution697.java

// https://leetcode-cn.com/problems/degree-of-an-array/
import java.util.HashMap;

public class Solution697 {
    public int findShortestSubArray(int[] nums) {
        HashMap<Integer, Integer[]> resMap = new HashMap<>();
        int maxRes = 0;
        int res = nums.length;

        for (int i = 0; i < nums.length; i++) {
            Integer[] resLst = resMap.getOrDefault(nums[i], new Integer[] { 0, i });
            resLst[0]++;

            resMap.put(nums[i], resLst);

            if (resMap.get(nums[i])[0] > maxRes) {
                maxRes = resMap.get(nums[i])[0];
                res = i - resMap.get(nums[i])[1] + 1;
            }

            if (resMap.get(nums[i])[0] == maxRes) {
                res = Math.min(res, i - resMap.get(nums[i])[1] + 1);
            }
        }

        return res;

    }

    public static void main(String[] args) {
        Solution697 s = new Solution697();
        int[] nums = { 1, 2, 2, 3, 1 };
        int res = s.findShortestSubArray(nums);
        System.out.println(res);
    }
}
